File: src/deep_blue_genome/coexpr/CytoscapeWriter.cpp
/*
 * Copyright (C) 2015 VIB/BEG/UGent - Tim Diels <timdiels.m@gmail.com>
 *
 * This file is part of Deep Blue Genome.
 *
 * Deep Blue Genome is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Deep Blue Genome is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Deep Blue Genome.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "CytoscapeWriter.h"
#include <iostream>
#include <fstream>
#include <functional>
#include <boost/range.hpp>
#include <boost/range/adaptors.hpp>
#include <boost/range/join.hpp>
#include <boost/range/algorithm_ext.hpp>
#include <boost/container/set.hpp>
#include <deep_blue_genome/common/database_all.h>
#include <deep_blue_genome/common/util.h>
#include <deep_blue_genome/common/writer/OrthologGroup.h>
#include <deep_blue_genome/coexpr/Baits.h>
#include <deep_blue_genome/coexpr/BaitCorrelations.h>
#include <deep_blue_genome/coexpr/OrthologGroupInfo.h>
#include <deep_blue_genome/coexpr/OrthologGroupInfos.h>
#include <deep_blue_genome/coexpr/BaitGroups.h>
#include <deep_blue_genome/util/printer.h>
#include <deep_blue_genome/util/functional.h>

using namespace DEEP_BLUE_GENOME;
using namespace DEEP_BLUE_GENOME::COMMON::WRITER;
using namespace boost::adaptors;
using namespace std;
using boost::container::flat_set;

namespace DEEP_BLUE_GENOME {
namespace COEXPR {

uint64_t Node::next_id = 1;

// TODO fantastic idea: header includes to third party libs used in a project -> use a precompiled header so we have less work on getting the includes right

CytoscapeWriter::CytoscapeWriter(std::string install_dir, const std::vector<Gene*>& baits, const std::vector<OrthologGroupInfo*>& neighbours, OrthologGroupInfos& groups)
:	install_dir(install_dir), baits(baits), neighbours(neighbours), groups(groups), network_name("network"), bait_orthologies_cached(false)
{
}

/**
 * Get orthology relations between baits
 */
std::vector<BaitBaitOrthRelation>& CytoscapeWriter::get_bait_orthology_relations() {
	if (!bait_orthologies_cached) {
		// get a set of ortholog groups of the baits
		flat_set<const OrthologGroupInfo*> bait_groups;
		for (auto bait : baits) {
			boost::insert(bait_groups, groups.get(*bait) | referenced);
		}

		// enumerate all possible pairs of baits of the same group
		for (auto group : bait_groups) {
			// filter out genes not present in our set of baits
			vector<const Gene*> genes;
			assert(group);
			for (auto gene : group->get().get_genes()) {
				if (contains(baits, gene)) {
					genes.emplace_back(gene);
				}
			}

			// output all pairs
			for (auto it = genes.begin(); it != genes.end(); it++) {
				for (auto it2 = it+1; it2 != genes.end(); it2++) {
					bait_orthologies.emplace_back(make_pair(*it, *it2));
					bait_orthologies.emplace_back(make_pair(*it2, *it));
				}
			}
		}
		erase_duplicates(bait_orthologies);
		bait_orthologies_cached = true;
	}

	return bait_orthologies;
}

/**
 * Write out a cytoscape network
 */
void CytoscapeWriter::write() {
	// write data to each file
	write_sif();
	write_node_attr();
	write_edge_attr();
	write_genes();

	// copy additional assets into the network's directory
	// copy_vizmap() TODO, without using boost copy_file (or try with a new header)
}

/**
 * Write node attributes
 *
 * Particularly, attributes of:
 * - bait nodes
 * - target nodes
 */
void CytoscapeWriter::write_node_attr() {
	ofstream out(network_name + ".node.attr");
	out.exceptions(ofstream::failbit | ofstream::badbit);
	auto fields = {"id", "label", "colour", "type", "bait_gene", "species", "families", "family", "correlating_genes_in_family"};
	out << intercalate("\t", fields) << "\n";

	write_node_attr_baits(out);
	write_node_attr_targets(out);
}

void CytoscapeWriter::write_node_attr_baits(ostream& out) {
	std::string colour = "#FFFFFF";
	for (auto bait : baits) {
		// fetch data
		auto&& gene = bait->get_name();
		auto&& node = bait_nodes[bait];
		auto&& label = gene;
		auto&& type = "bait node";
		auto&& species = bait->get_gene_collection().get_species();
		auto&& formatted_families = intercalate(". ", groups.get(*bait) | transformed(bind(&OrthologGroupInfo::get, std::placeholders::_1)) | transformed(OrthologGroupWriter::format_long_id));

		// write line
		out << intercalate_("\t", node.get_cytoscape_id(), label, colour, type, gene, species, formatted_families, "", "") << "\n";
	}
}

void CytoscapeWriter::write_node_attr_targets(ostream& out) {
	// have each neigh figure out what its bait group is
	BaitGroups bait_groups;
	for (auto neigh : neighbours) {
		neigh->init_bait_group(bait_groups);
	}

	// assign colours to bait groups
	std::default_random_engine generator;
	std::uniform_int_distribution<int> distribution(0, 0x00FFFFFF);
	for (auto& p : bait_groups) {
		auto& group = p.second;

		ostringstream str;
		int colour = distribution(generator);
		str << "#";
		str.width(6);
		str.fill('0');
		str << std::hex << colour;

		group.set_colour(str.str());
	}

	// output targets (= family nodes)
	for (auto&& neigh : neighbours) {
		// gene_names
		auto&& get_name = make_function([](const Gene* g) { // TODO could use std bind
			return g->get_name();
		});
		auto&& corr_genes_in_fam = intercalate(", ", neigh->get_correlating_genes() | transformed(get_name));

		// other
		auto&& label = corr_genes_in_fam;
		auto&& node = target_nodes[neigh];
		auto&& colour = neigh->get_bait_group().get_colour();
		auto&& type = "family node";
		auto&& formatted_family = OrthologGroupWriter::format_long_id(neigh->get());

		// write line
		out << intercalate_("\t", node.get_cytoscape_id(), label, colour, type, "", "", "", formatted_family, corr_genes_in_fam) << "\n";
	}
}

/**
 * Write yaml file with info of each gene
 */
void CytoscapeWriter::write_genes() {
	YAML::Node root;

	// baits to yaml nodes
	for (auto gene : baits) {
		root["genes"].push_back(get_bait_node(*gene));
	}

	// collect targets
	flat_set<const Gene*> targets;
	for (auto neigh : neighbours) {
		boost::insert(targets, neigh->get_correlating_genes());
	}

	// targets to yaml nodes
	for (auto gene : targets) {
		for (auto&& family_info : groups.get(*gene)) {
			root["genes"].push_back(get_family_node(family_info));
		}
	}

	// Write to file
	ofstream out(network_name + "_genes.yaml");
	out.exceptions(ofstream::failbit | ofstream::badbit);
	out << YAML::Dump(root);
}

YAML::Node CytoscapeWriter::get_bait_node(const Gene& gene) {
	auto&& family_infos = groups.get(gene);

	YAML::Node gene_;
	gene_["id"] = gene.get_name(); // matches gene ids used in the node attr file (genes column)  TODO shouldn't this match the node id instead?
	// gene_["go_terms"] = TODO;
	gene_["is_bait"] = true;

	// families
	for (auto&& family : family_infos) {
		gene_["families"].push_back(write_yaml(family.get())); // TODO now returning a sequence, will need to adjust the java plugin
	}

	// orthologs
	flat_set<std::string> orthologs;
	for (auto&& family : family_infos) {
		for (auto&& g : family.get_correlating_genes()) {
			if (g != &gene) {
				orthologs.insert(g->get_name());
			}
		}
	}
	gene_["orthologs"] = std::vector<std::string>(orthologs.begin(), orthologs.end());

	return gene_;
}

YAML::Node CytoscapeWriter::get_family_node(const OrthologGroupInfo& group) {
	YAML::Node node;
	node["id"] = target_nodes[&group].get_cytoscape_id(); // matches gene ids used in the node attr file (genes column)
	// gene_["go_terms"] = TODO;
	node["is_bait"] = false;

	// baits
	for (auto bait_correlation : group.get_bait_correlations()) {
		YAML::Node bait;
		bait["node_id"] = bait_nodes[&bait_correlation.get_bait()].get_cytoscape_id();
		bait["r_value"] = bait_correlation.get_max_correlation();
		node["baits"].push_back(bait);
	}

	return node;
}


/**
 * Write node relations
 *
 * Particularly:
 * - target->bait correlation relations
 * - bait <-> bait orthology relations
 */
void CytoscapeWriter::write_sif() {
	ofstream out(network_name + ".sif");
	out.exceptions(ofstream::failbit | ofstream::badbit);

	// target -> bait correlation
	for (auto& neigh : neighbours) {
		auto&& delimiter = "\t";
		auto&& prefix = intercalate_(delimiter, target_nodes[neigh].get_cytoscape_id(), "cor");
		if (!neigh->get_bait_correlations().empty()) {
			auto get_name = make_function([this](const BaitCorrelations& bait_correlation) {
				return bait_nodes[&bait_correlation.get_bait()].get_cytoscape_id();
			});
			auto&& bait_nodes_ = intercalate(delimiter, neigh->get_bait_correlations() | transformed(get_name));
			out << intercalate_(delimiter, prefix, bait_nodes_) << "\n";
		}
	}

	// bait <-> bait orthology
	for (auto& p : get_bait_orthology_relations()) {
		out << bait_nodes[p.first].get_cytoscape_id() << "\thom\t" << bait_nodes[p.second].get_cytoscape_id() << "\n";
		// TODO we are printing pairs per line, not: src rel dst1 dst2 .. dstN. Does that read stuff fine?
	}
}

/**
 * Write same node relations as sif, but in more detail
 */
void CytoscapeWriter::write_edge_attr() {
	ofstream out(network_name + ".edge.attr");
	out.exceptions(ofstream::failbit | ofstream::badbit);
	out << "edge\tr_value\n";

	// target -> bait correlation
	for (auto&& neigh : neighbours) {
		if (!neigh->get_bait_correlations().empty()) {
			for (auto& bait_correlation : neigh->get_bait_correlations()) {
				out << target_nodes[neigh].get_cytoscape_id() << " (cor) " << bait_nodes[&bait_correlation.get_bait()].get_cytoscape_id() << "\t" << bait_correlation.get_max_correlation() << "\n";
			}
		}
	}

	// bait <-> bait orthology
	for (auto& p : get_bait_orthology_relations()) {
		out << bait_nodes[p.first].get_cytoscape_id() << " (hom) " << bait_nodes[p.second].get_cytoscape_id() << "\tNA\n";
	}
}

}} // end namespace

File: python3/src/deep_blue_genome/coexpnetviz/cytoscapewriter.py
# Copyright (C) 2015 VIB/BEG/UGent - Tim Diels <timdiels.m@gmail.com>
# 
# This file is part of Deep Blue Genome.
# 
# Deep Blue Genome is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# Deep Blue Genome is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public License
# along with Deep Blue Genome.  If not, see <http://www.gnu.org/licenses/>.

import pandas as pd
import numpy as np
from deep_blue_genome.core.util import series_invert, get_distinct_colours
from deep_blue_genome.pkg_util import get_data_file


# TODO consider one of the formats that allows specifying the network, edges with attributes and nodes with attributes all at once. Perhaps even with a style.
# e.g. XGMLL from http://wiki.cytoscape.org/Cytoscape_3/UserManual#Cytoscape_3.2BAC8-UserManual.2BAC8-Network_Formats.XGMML_Format
class CytoscapeWriter(object):
    def __init__(self, network):
        self._network = network
        
    def write(self):
        # assign node ids to baits: pd.Series(index=(gene : str), data=(id : int))
        self._bait_nodes = self._network.baits.copy()
        self._bait_nodes.index.name = 'id'
        self._bait_nodes = series_invert(self._bait_nodes)
        
        # assign node ids to family nodes that have a non-nan family
        next_id = self._bait_nodes.max() + 1
        family_ids = self._network.correlations['family']
        family_ids.dropna(inplace=True)
        family_ids.drop_duplicates(inplace=True)
        family_ids = family_ids.to_frame()
        family_ids['family_id'] = range(next_id, next_id+len(family_ids))
        next_id += len(family_ids)
        family_ids.set_index('family', inplace=True)
        self._correlations = self._network.correlations.set_index('family').join(family_ids)
        self._correlations.reset_index(inplace=True)
        del family_ids
        
        # assign node ids to family nodes with nan family
        mask = self._correlations['family_id'].isnull()
        self._correlations.loc[mask, 'family_id'] = range(next_id, next_id + mask.sum())
        
        # throw in a bait_id column
        # self._correlations : pd.DataFrame(columns=[(bait : str), (family : str), (family_gene : str), (correlation : float), (family_id : int), (bait_id : int)])
        self._correlations.set_index('bait', inplace=True)
        self._correlations = self._correlations.join(self._bait_nodes)
        self._correlations.reset_index(inplace=True)
        self._correlations.rename(columns={'index': 'bait', 'id': 'bait_id'}, inplace=True)
        
        # group correlation and family_gene
        # self._correlations : pd.DataFrame(columns=[(bait : str), (family : str), (family_id : int), (bait_id : int), (correlating_genes_in_family : str), (max_correlation : float)]) 
        max_correlations = self._correlations.groupby(['family_id', 'bait_id'])['correlation'].max()
        max_correlations.name = 'max_correlation'
        self._correlations.drop('correlation', axis=1, inplace=True)
        self._correlations.drop_duplicates('family_gene', inplace=True)
        
        correlating_genes_in_family = self._correlations.groupby('family_id')['family_gene'].apply(lambda x: ', '.join(x.tolist()))
        correlating_genes_in_family.name = 'correlating_genes_in_family'
        self._correlations.drop('family_gene', axis=1, inplace=True)
        self._correlations.drop_duplicates(inplace=True)
        
        self._correlations = self._correlations.join(correlating_genes_in_family, on='family_id')
        self._correlations = self._correlations.join(max_correlations, on=['family_id', 'bait_id'])
        
        # bait_families: pd.Series(index=(family : str), data=(id : int))
        bait_families = self._network.gene_families.map(self._bait_nodes)
        bait_families.dropna(inplace=True)
        bait_families.name = self._bait_nodes.name
        
        # Write it
        self.write_sif(bait_families)
        self.write_node_attr(bait_families)
        self.write_edge_attr()
        
        # Copy additional files
        for file in [get_data_file('coexpnetviz/coexpnetviz_style.xml'), get_data_file('coexpnetviz/README.txt')]:
            file.copy('.')
    
    def write_sif(self, bait_families):
        # correlation edges (without attribute data)
        correlation_edges = self._correlations.copy()
        correlation_edges['type'] = 'cor'
        correlation_edges = correlation_edges[['family_id', 'type', 'bait_id']]
        
        # homology edges
        homology_edges = pd.concat([bait_families]*2, axis=1)
        homology_edges.columns = ['id1', 'id2']
        homology_edges = homology_edges[homology_edges['id1'] >  homology_edges['id2']]  # drop self relations
        homology_edges['type'] = 'hom'
        homology_edges = homology_edges.reindex(columns='id1 type id2'.split())
        
        # sif
        bait_nodes = self._bait_nodes.to_frame(0)
        correlation_edges.columns = range(3)
        homology_edges.columns = range(3)
        
        sif = pd.concat([bait_nodes, correlation_edges, homology_edges], ignore_index=True)
        sif[[0,2]] = sif[[0,2]].applymap(self._format_node_id)
        sif.to_csv('{}.sif'.format(self._network.name), sep='\t', header=False, index=False)
    
    def write_node_attr(self, bait_families):
        ###################
        # bait node attrs
        bait_node_attrs = self._bait_nodes.reset_index()
        bait_families = bait_families.reset_index().groupby('id')['family'].agg(lambda x: ', '.join(x.tolist()))
        bait_families.name = 'families'
        bait_node_attrs = bait_node_attrs.join(bait_families, on='id')
        bait_node_attrs.rename(columns={'gene': 'bait_gene'}, inplace=True)
        bait_node_attrs['label'] = bait_node_attrs['bait_gene']
        bait_node_attrs['colour'] = '#FFFFFF'
        bait_node_attrs['type'] = 'bait node'
        bait_node_attrs['partition_id'] = hash(frozenset())
        
        ###################
        # Family node attrs
        
        # partitions: pd.Series(index=(family_id : int), data=(partition_id : int))
        partitions = self._correlations.groupby('family_id')['bait_id'].agg(lambda x: hash(frozenset(x.tolist())))
        partitions.name = 'partition_id'
        
        # colours assigned to partitions: pd.Series(index=(partition_id : int), data=(colour : str))
        colours = partitions.drop_duplicates()
        colours = pd.DataFrame(get_distinct_colours(len(colours)), index=colours)
        def to_hex(x):
            return '#' + ''.join((x*255).round().apply(lambda y: '{:02x}'.format(int(y))))
            #return ''.join(['#'] + list(map(lambda x: .format(min(255, round(x*255))), tup)))
        colours = colours.apply(to_hex, axis=1)
        colours = colours.reindex(np.random.permutation(colours.index))  # shuffle the colours so distinct colours are less likely to be put next to each other
        colours.name = 'colour'
        
        # family_colours : pd.Series(index=(family_id : int), data=(colour : str))
        family_colours = partitions.to_frame().join(colours, on='partition_id')
        
        # Family node attrs
        family_node_attrs = self._correlations[['family_id', 'family', 'correlating_genes_in_family']].set_index('family_id')
        family_node_attrs['type'] = 'family node'
        family_node_attrs['label'] = family_node_attrs['correlating_genes_in_family'] 
        family_node_attrs = family_node_attrs.join(family_colours)
        family_node_attrs.index.name = 'id'
        family_node_attrs.reset_index(inplace=True)
        
        #######################
        # concat
        nodes = pd.concat([bait_node_attrs, family_node_attrs], ignore_index=True)
        nodes = nodes.reindex(columns='id label colour type bait_gene species families family correlating_genes_in_family partition_id'.split())
        nodes['id'] = nodes['id'].apply(self._format_node_id)
        nodes.to_csv('{}.node.attr'.format(self._network.name), sep='\t', index=False)
        
    def write_edge_attr(self):
        correlation_edges = self._correlations[['family_id', 'bait_id']]
        correlation_edges = correlation_edges.apply(lambda x: '{} (cor) {}'.format(*map(self._format_node_id, x.tolist())), axis=1).to_frame('edge')
        correlation_edges['max_correlation'] = self._correlations['max_correlation']
        correlation_edges.to_csv('{}.edge.attr'.format(self._network.name), sep='\t', index=False)
        
    def _format_node_id(self, node_id):
        if np.isnan(node_id):
            return ''
        else:
            return 'n{}'.format(int(node_id))

    
    
